worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala

```diff
@@ -526,4 +530,278 @@ class PushDataHandler extends BaseMessageHandler with Logging {
+        }
+        case Type.REGION_START => {
+          fileWriter.regionStart(
+            message.asInstanceOf[RegionStart].currentRegionIndex,
+            message.asInstanceOf[RegionStart].isBroadcast)
+        }
+        case Type.REGION_FINISH => {
+          fileWriter.regionFinish()
+        }
+      }
+      // for master, send data to slave
+      if (location.getPeer != null && isMaster) {
+        // to do replica
+        wrappedCallback.onSuccess(ByteBuffer.wrap(Array[Byte]()))
+      } else {
+        wrappedCallback.onSuccess(ByteBuffer.wrap(Array[Byte]()))
+      }
+    } catch {
+      case t: Throwable =>
+        callback.onFailure(new Exception(s"$messageType failed", t))
+    }
+  }
+
+  class WrappedRpcResponseCallback(
+      messageType: Message.Type,
+      isMaster: Boolean,
+      requestId: Long,
+      softSplit: AtomicBoolean,
+      location: PartitionLocation,
+      workerSourceTime: String,
+      callback: RpcResponseCallback)
+    extends RpcResponseCallback {
+    override def onSuccess(response: ByteBuffer): Unit = {
+      if (isMaster) {
+        workerSource.stopTimer(workerSourceTime, s"$requestId")
+        if (response.remaining() > 0) {
+          val resp = ByteBuffer.allocate(response.remaining())
+          resp.put(response)
+          resp.flip()
+          callback.onSuccess(resp)
+        } else if (softSplit != null && softSplit.get()) {
+          callback.onSuccess(ByteBuffer.wrap(Array[Byte](StatusCode.SOFT_SPLIT.getValue)))
+        } else {
+          callback.onSuccess(response)
+        }
+      } else {
+        workerSource.stopTimer(workerSourceTime, s"$requestId")
+        callback.onSuccess(response)
+      }
+    }
+
+    override def onFailure(e: Throwable): Unit = {
+      if (location != null) {
+        logError(s"[handle$messageType.onFailure] partitionLocation: $location")
+      }
+      messageType match {
+        case Type.PUSH_DATA_HAND_SHAKE =>
+          workerSource.incCounter(WorkerSource.PushDataHandshakeFailCount)
+          callback.onFailure(new Exception(
+            StatusCode.PUSH_DATA_HANDSHAKE_FAIL_SLAVE.getMessage,
+            e))
+        case Type.REGION_START =>
+          workerSource.incCounter(WorkerSource.RegionStartFailCount)
+          callback.onFailure(new Exception(StatusCode.REGION_START_FAIL_SLAVE.getMessage, e))
+        case Type.REGION_FINISH =>
+          workerSource.incCounter(WorkerSource.RegionFinishFailCount)
+          callback.onFailure(new Exception(StatusCode.REGION_FINISH_FAIL_SLAVE.getMessage, e))
+        case _ =>
+          workerSource.incCounter(WorkerSource.PushDataFailCount)
+          callback.onFailure(new Exception(StatusCode.PUSH_DATA_FAIL_SLAVE.getMessage, e))
+      }
+    }
+  }
+
+  private def checkLocationNull(
+      messageType: Message.Type,
+      shuffleKey: String,
+      partitionUniqueId: String,
+      body: ByteBuf,
```

Comment:
Looks like the body is always null, maybe this variable can be removed.

Reply:
yes, for message handshake/regionstart/regionfinish,  the body is always null, but for message pushdata,  it will be not null. in the future , i will refactor pushdata's logic to use thie method